### PR TITLE
Make tikv-server fully generic over kv engines

### PR DESCRIPTION
**Pull request by brson**
### What problem does this PR solve?

https://github.com/tikv/tikv/issues/6402
https://github.com/tikv/tikv/issues/10528

Problem Summary:

Make tikv-server startup generic over engines.

### What is changed and how it works?

This is a rewrite of previous pr https://github.com/tikv/tikv/pull/10529. This iteration has a cleaner history, but ends at basically the same place.

The previous patch bitrotted badly after the addition of FlowController, which is rocks-specific. This patch makes it so that FlowController should work if not set up. It does this by using the FlowController::empty ctor, which is documented as "test only", but would seem to create a basically-working FlowController.

The rest of the series is moving routines that are engine-specific into a new trait, CreateKvEngine.

After this patch, a second engine can be added in a single commit. I have not done this yet to avoid the question of whether and how to launch PanicEngine, which as a feature is useless, but is the only other in-tree implementation of KvEngine.

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by brson** [comment]
I intend a future patch to convert the raft engine setup code to use a CreateRaftEngine trait, like the CreateKvEngine crate in this PR. I think it will be a simplification.

**Comment by brson** [comment]
I went ahead and pushed the commits to add the CreateRaftEngine trait. This makes the setup routine a generic function instead of a macro. So after this patch, kv engines need to implement CreateKvEngine and raft engines CreateRaftEngine.

**Comment by brson** [comment]
With the addition of CreateRaftEngine, the patch to enable PanicEngine for both KvEngine and RaftEngine is not so ugly, so I pushed that as well.

After this patch it is possible to launch with PanicEngine as either the KV engine or the raft engine.

File: components/server/src/server.rs [hunk 1]
```diff
@@ -1218,38 +1207,76 @@ impl<ER: RaftEngine> TiKVServer<ER> {
...
+trait CreateKvEngine<ER>: KvEngine
+where
+    ER: RaftEngine,
+{
+    fn create_kv_engine(
+        config: &TiKvConfig,
+        region_info_accessor: &RegionInfoAccessor,
+        store_path: &Path,
+        router: &RaftRouter<Self, ER>,
+        env: Arc<engine_rocks::raw::Env>,
```
Comment (Connor1996) [inline components/server/src/server.rs:1222]: it's still specific to rocksDB

Comment (brson) [reply components/server/src/server.rs:1222]: Yes, that's right. There are several features that remain rocks-specific for now, including the debug and backup services and this encryption configuration. This patch enables non-rocks engines to ignore those features and still boot.

Comment (brson) [reply components/server/src/server.rs:1222]: It's not in this series, but a patch that introduces other engines implements some of these CreateKvEngine methods as no-ops.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by brson** [comment]
/test

please test?

**Comment by zhouqiang-cl** [comment]
@Mini256 please take a look at the invaid-commit-message or give a guide how to handle with invalid commit message

**Comment by Mini256** [comment]
> [FORMAT CHECKER NOTIFICATION]
> 
> Please provide the associated issue number in the commit messages, for example: `close #12345`, or `ref #12345`.

@brson, Hi, because the [vote-0150](https://github.com/tikv/community/blob/master/votes/0150-pr-issue-association.md) passed, the process of submitting PR has changed some changes: We need to provide the issue number associated with PR in the commit message.

To pass the format checker, you can attach the issue number in the previous commit message through the `git commit --amend` command, If the associated issue will be closed by merged PR, you can use `close #123`, if you just reference the related issue, you can use `ref #456`, for example:

```
server: some message
close #123
```

The format checker will ensure that at least one commit‘s message carries the issue number before the Pull Request is merged. 

Before a new Pull Request is submitted, we can fill in the `close # <issue_number>` or `ref # <issue_number>` by `git commit -s -a` command in advance. For example:

image

**Comment by kennytm** [comment]
/test

https://ci.pingcap.net/blue/organizations/jenkins/tikv_ghpr_test/detail/tikv_ghpr_test/11840/pipeline

`cases::test_gc_worker::test_collect_applying_locks` (#8691?)

**Comment by brson** [comment]
Thanks @zhouqiang-cl . I'll look into it.

**Pull request merged**